### PR TITLE
feat: add erc20votes voting strategies

### LINKED ISSUE #478: integrate erc20votes
https://github.com/OpenZeppelin/cairo-contracts/pull/685

### PR DESCRIPTION
**Pull request by pscott**
Added a branch of OZ as dependency, can move to the actual OZ repo once the PR gets merged.
Also, had to copy the ERC20 Votes Preset because of [a scarb bug](https://github.com/software-mansion/scarb/discussions/568#discussioncomment-6742412).

Closes #478

### CODE REVIEW

File: starknet/src/voting_strategies/erc20_votes.cairo [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+            voter: UserAddress,
+            params: Array<felt252>,
+            user_params: Array<felt252>,
+        ) -> u256 {
+            // Cast voter address to a Starknet address
+            // Will revert if the address is not a Starknet address
+            let voter = voter.to_starknet_address();
+
+            // Get the ERC20 contract address from the params array
+            let mut params_span = params.span();
```
Comment (Orland0x) [inline starknet/src/voting_strategies/erc20_votes.cairo]: can we just shadow the var in these instances? like `let mut params = params`

File: starknet/src/voting_strategies/erc20_votes.cairo [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+            let erc20_contract_address = Serde::<ContractAddress>::deserialize(ref params_span)
+                .unwrap();
+
+            let erc20 = IVotesDispatcher { contract_address: erc20_contract_address,  };
+
+            erc20.get_past_votes(voter, timestamp.into()) // TODO; verify - 1
```
Comment (Orland0x) [inline starknet/src/voting_strategies/erc20_votes.cairo]: The space should enforce its never called with the current timestamp so dont need to check again here

File: starknet/src/tests/voting_strategies/erc20_votes.cairo [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+
+        testing::set_block_timestamp(current + config.max_voting_duration);
+
+        // Execute proposal
+        space.execute(1_u256, vanilla_execution_strategy.params);
+    }
+
+    #[test]
+    #[available_gas(1000000000)]
+    #[should_panic]
```
Comment (Orland0x) [inline starknet/src/tests/voting_strategies/erc20_votes.cairo]: we should check panic for the specific error expected , easy to miss bugs otherwise

**Review by Orland0x — COMMENTED** [review]
Looks good

File: starknet/src/voting_strategies/erc20_votes.cairo (hunk 1 above)
Comment (pscott) [reply starknet/src/voting_strategies/erc20_votes.cairo]: 091aaf360dfa3efd269cd3882a4a88d06058747c

File: starknet/src/voting_strategies/erc20_votes.cairo (hunk 2 above)
Comment (pscott) [reply starknet/src/voting_strategies/erc20_votes.cairo]: c9d4a6ef44f30f003981e4b2527ed07bf374320a

File: starknet/src/tests/voting_strategies/erc20_votes.cairo (hunk 3 above)
Comment (pscott) [reply starknet/src/tests/voting_strategies/erc20_votes.cairo]: you are 100% correct, and one of the test was indeed not panicking for the correct reason :p

47ad5cc884dace269e852ac3f033b8560bbcc0e5

**Review by Orland0x — APPROVED** [review]
uTACK

**Pull request merged**